Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py

```diff
@@ -88,72 +148,206 @@ def PyExec(self):
-            if isinstance(mtd['__cal_tmp'], IEventWorkspace):
-                Integration(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            CopyInstrumentParameters(data, '__cal_tmp', EnableLogging=False)
-            ConvertSpectrumAxis(InputWorkspace='__cal_tmp', Target=target, EFixed=eFixed, OutputWorkspace='__cal_tmp', EnableLogging=False)
-            Transpose(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            ResampleX(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', XMin=xMin, XMax=xMax, NumberBins=numberBins,
-                      EnableLogging=False)
-            Divide(LHSWorkspace=outWS, RHSWorkspace='__cal_tmp', OutputWorkspace=outWS, EnableLogging=False)
-            if normaliseBy == "Monitor":
-                cal_scale = cal.run().getProtonCharge()
-            elif normaliseBy == "Time":
-                cal_scale = cal.run().getLogData('duration').value
-
-        Scale(InputWorkspace=outWS, OutputWorkspace=outWS, Factor=cal_scale/data_scale, EnableLogging=False)
-
-        if bkg is not None:
-            ExtractUnmaskedSpectra(InputWorkspace=bkg, MaskWorkspace='__mask_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            if isinstance(mtd['__bkg_tmp'], IEventWorkspace):
-                Integration(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            CopyInstrumentParameters(data, '__bkg_tmp', EnableLogging=False)
-            ConvertSpectrumAxis(InputWorkspace='__bkg_tmp', Target=target, EFixed=eFixed, OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            Transpose(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            ResampleX(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', XMin=xMin, XMax=xMax, NumberBins=numberBins,
-                      EnableLogging=False)
+        _get_scale = lambda x: {
+            None: 1,
+            "Monitor": x.run().getProtonCharge(),
+            "Time": x.run().getLogData("duration").value,
+        }[normaliseBy]
+
+        # NOTE:
+        # StringArrayProperty cannot be optional, so the background can only be passed in as a string
+        # or a list, which will be manually unpacked here
+        try:
+            bkg = bkg.split(",") if bkg is not None else bkg
+        except:
+            _bg = AnalysisDataService.retrieve(
+                bkg[0]
+            )  # invalid background ws will be detected here
+
+        # NOTE:
+        # Due to range difference among incoming spectra, a common bin para is needed
+        # such that all data can be binned exactly the same way.
+
+        # Step_1: Find the global bin from all spectrum
+        _xMin = 1e16
+        _xMax = -1e16
+        # BEGIN_FOR: located_global_xMin&xMax
+        for n, _wsn in enumerate(data):
```

Comment:
It is better to point out that data is a list of workspace name.

Reply:
All new methods have explicit name for their corresponding data source (input, calibration, background)